sympy/core/numbers.py

```diff
@@ -207,6 +211,149 @@ def igcd(*args):
     return a
 
 
+def igcd_lehmer(a, b):
+    """Computes greatest common divisor of two integers.
+
+    Euclid's algorithm for the computation of the greatest
+    common divisor  gcd(a, b)  of two (positive) integers
+    a and b is based on the division identity
+        a = q*b + r,
+    where the quotient  q  and the remainder  r  are integers
+    and  0 <= r < b. Then each common divisor of  a  and  b
+    divides  r, and it follows that  gcd(a, b) == gcd(b, r).
+    The algorithm works by constructing the sequence
+    r0, r1, r2, ..., where  r0 = a, r1 = b,  and each  rn
+    is the remainder from the division of the two preceding
+    elements.
+
+    In Python, q = a // b  and  r = a % b  are obtained by the
+    floor division and the remainder operations, respectively.
+    These are the most expensive arithmetic operations, especially
+    for large  a  and  b.
+
+    Lehmer's algorithm is based on the observation that the quotients
+    qn = r(n-1) // rn  are in general small integers even
+    when  a  and  b  are very large. Hence the quotients can be
+    usually determined from a relatively small number of most
+    significant bits.
+
+    The efficiency of the algorithm is further enhanced by not
+    computing each long remainder in Euclid's sequence. The remainders
+    are linear combinations of  a  and  b  with integer coefficients
+    derived from the quotients. The coefficients can be computed
+    as far as the quotients can be determined from the chosen
+    most significant parts of  a  and  b. Only then a new pair of
+    consecutive remainders is computed and the algorithm starts
+    anew with this pair.
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Lehmer%27s_GCD_algorithm
+
+    """
+    a, b = abs(as_int(a)), abs(as_int(b))
+    if a < b:
+        a, b = b, a
+
+    # The algorithm works by using one or two digit division
+    # whenever possible. The outer loop will replace the
+    # pair (a, b) with a pair of shorter consecutive elements
+    # of the Euclidean gcd sequence until a and b
+    # fit into two Python (long) int digits.
+    nbits = 2*int_info.bits_per_digit
+
+    while a.bit_length() > nbits and b != 0:
+        # Quotients are mostly small integers that can
+        # be determined from most significant bits.
+        n = a.bit_length() - nbits
+        x, y = a >> n, b >> n  # most significant bits
```

Comment:
Might be worth it to cast `x` and `y` to `int` here for Python 2, so that they are indeed native integers. Sadly, for Python 3, all ints are long ints.